File: path_sum_2_lc113.py
"""
Name: Srinidhi 
Did it run on LC: Yes

Time Complexity - O(n)
Space Complexity: O(n) -for the new path created 

Logic: As we do a DFS from root to leaf, keep subtracting value of the node we vsited
When we reach root, check if the remaining sum is equal top root value
and check for root condition.

Maintain a list for path and a list for result.

"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if root is None:
            return []
        res = []
        
        def dfsHelper(node,path,res,curr):
            if node is None:
                return
            
            path.append(node.val)
            
            
            if curr == node.val and node.left is None and node.right is None:
                res.append(list(path))
            
            else:
                dfsHelper(node.left,path,res,curr-node.val)
                dfsHelper(node.right,path,res,curr-node.val)
            path.pop()
        dfsHelper(root,[],res,sum)
        return res

File: PathSum2-LC113.java
/**
T: O(n) as we are vising each node
S: O(d) under the hood stack, max depth of the tree 
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int sum = 0;
    List<List<Integer>> result;
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        result = new ArrayList<List<Integer>>();
        if(root == null)
            return result;
        
        sum = targetSum;
        dfs(root, 0, new ArrayList<>());  
        return result;
    }
    
    public void dfs(TreeNode root, int cursum, List<Integer> path){
        //base case
        if(root == null)
            return ;
        //logic
        path.add(root.val);
        cursum += root.val;

        //check it the sum is equal to target sum and it is leaf node
        if(cursum == sum && root.left == null && root.right == null){
            result.add(new ArrayList<>(path));
            //return if we found the path
            path.remove(path.size() - 1);
            return;
        }
            
        dfs(root.left, cursum, path);        
        dfs(root.right, cursum, path);
        path.remove(path.size() - 1);
    }
}